### PR TITLE
Adopt Latest PHC Version

### PR DESCRIPTION
**Pull request by fire-at-will**
This PR updates the KMP SDK to consume the latest changes in PHC 13.0.1. Once it is merged, we should be ready for a beta 3 release.

### Description
Specifically, this PR:
- Bumps required OS versions and PHC version
   - iOS 13
   - PHC 13.0.1
- Introduces a new `StoreKitVersion` enum
- Changes the PurchasesAreCompletedBy enum to now be a sealed class, requiring a `storeKitVersion` to be provided when the value is set to `MyApp`
- Adds `storeKitVersion` to the PurchasesConfiguration object
- Adds the `recordPurchase` function to `Purchases`.
   - Changes `StoreTransaction` to be a class on iOS. For now, it remains a typealias on Android due to some conflicts in `purchases-ui`
- Introduces unit tests for the first time! 😄
- Runs the unit tests in CI for pull requests and on the main branch.
- Updates the apiTester and  apiDump files

### CODE REVIEW

File: core/src/iosMain/kotlin/com/revenuecat/purchases/kmp/Purchases.ios.kt
Comment (JayShortway) [inline core/src/iosMain/kotlin/com/revenuecat/purchases/kmp/Purchases.ios.kt:23]: There are a bunch of code style changes in this file. We don't really have one setup, so for now we're using the Android Studio default. Do you have it set to something custom?

File: apiTester/src/androidMain/kotlin/com/revenuecat/purchases/kmp/apitester/PurchasesAreCompletedByAPI.kt
Comment (fire-at-will) [reply apiTester/src/androidMain/kotlin/com/revenuecat/purchases/kmp/apitester/PurchasesAreCompletedByAPI.kt]: Love this! I've addressed this in commit https://github.com/RevenueCat/purchases-kmp/pull/139/commits/f1c166cecb84c31561f5b3180cee2bfa37042b45.

File: core/src/androidMain/kotlin/com/revenuecat/purchases/kmp/Purchases.android.kt
Comment (fire-at-will) [reply core/src/androidMain/kotlin/com/revenuecat/purchases/kmp/Purchases.android.kt]: I think that calling onError is definitely the better developer experience, since the existence of the `onError` parameter sort of implies that the function won't throw. I went ahead and updated this to call `onError` in commit https://github.com/RevenueCat/purchases-kmp/pull/139/commits/cda83262d8d2f1ed7657c97ecd76a131cdc81cd4.

It looks like we're currently calling `error()` in other places in the SDK when a function isn't supported, so this does introduce some inconsistency in how we handle unsupported functions. I'm okay shipping this either way (we can revert https://github.com/RevenueCat/purchases-kmp/pull/139/commits/cda83262d8d2f1ed7657c97ecd76a131cdc81cd4 if we don't want to keep it), but we should probably consider standardizing how we handle unsupported functions before the 1.0 release.

References:
- [Purchases.android.getPromotionalOffer](https://github.com/RevenueCat/purchases-kmp/blob/bdbadfe3d6deb3b0dacb06b502990c0b6de32687/core/src/androidMain/kotlin/com/revenuecat/purchases/kmp/Purchases.android.kt#L163) calls `error()` 
- [This Purchases.ios.purchase function](https://github.com/RevenueCat/purchases-kmp/blob/bdbadfe3d6deb3b0dacb06b502990c0b6de32687/core/src/iosMain/kotlin/com/revenuecat/purchases/kmp/Purchases.ios.kt#L210) calls `error`
- [Purchases.ios.syncAmazonPurchase](https://github.com/RevenueCat/purchases-kmp/blob/bdbadfe3d6deb3b0dacb06b502990c0b6de32687/core/src/iosMain/kotlin/com/revenuecat/purchases/kmp/Purchases.ios.kt#L126) is a no-op and doesn't call `error()` or throw

File: core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesAreCompletedBy.kt
Comment (fire-at-will) [reply core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesAreCompletedBy.kt]: Great idea! No reason to have it as a class if it can be an interface. Updated in https://github.com/RevenueCat/purchases-kmp/pull/139/commits/b0cac80e7c4133a2ce8952e215a0a3fbe222d80c

File: core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt
Comment (fire-at-will) [reply core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt:48]: > Question for my understanding: why do we have the separate storeKitVersion config option? Why not exclusively use the version set by PurchasesAreCompletedBy?

This is a _great_ question! 😄 The reason has to do with some implementation details with the StoreKit and PurchasesAreCompletedBy changes in the 5.X iOS SDK.

An optional `storeKit: StoreKitVersion` parameter was introduced to allow developers to specify which StoreKit version the SDK uses to process purchases. It's an optional parameter that defaults to StoreKitVersion.DEFAULT (which points to SK2) because ideally developers don't need to think about which SK version they're using (that's why they're using RC after all 😄).

On the other hand, we had to write a completely separate implementation of `PurchasesAreCompletedByMyApp` for SK2. When the SDK is initialized with `PurchasesAreCompletedBy==MyApp`, we need to know which `PurchasesAreCompletedByMyApp` implementation to wire up - the SK1 or SK2 implementation, so that's why providing a `StoreKitVersion` is mandatory when using `PurchasesAreCompletedByMyApp`. We're okay asking for a StoreKit version here because any developer who is already processing their own purchases should know the difference between SK1 and SK2, and which one they're using.

The reason that we can't exclusively use the version set by `PurchasesAreCompletedBy` is that we wanted people to be able to specify the StoreKit version without having to think about what `PurchasesAreCompletedBy` was. In retrospect, we might could have simplified this by adding `storeKitVersion` to `PurchasesAreCompletedBy.RevenueCat` and forgoing the `storeKitVersion` parameter entirely, but this is the API we settled on for this release 😅

File: core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesAreCompletedBy.kt
Comment (fire-at-will) [reply core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesAreCompletedBy.kt]: Looking at this function again, it looks like it was built earlier in this PR's development, but it turns out that it wasn't needed, and it's currently only used in the tests! 😅

I've removed it in https://github.com/RevenueCat/purchases-kmp/pull/139/commits/d52bf4739e676fc52a5f0965d477e1f6a361eb1c 😄

File: core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt
Comment (fire-at-will) [reply core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt]: I've moved this from using `println` to `LogHandler` in https://github.com/RevenueCat/purchases-kmp/pull/139/commits/85fc14d4fd5eebd11a44289000dd1eaa5497c738. I used `[Purchases]` as the tag, do you know if there's a more appropriate value to use instead?

File: core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt
Comment (fire-at-will) [reply core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt]: The reason we have `DEFAULT` in the `StoreKitVersion` is to allow developers to explicitly tell our SDK to pick the most appropriate version of StoreKit to use given the app's runtime environment. We have the `DEFAULT` value in the native and other hybrid SDKs, so I'd lean towards keeping it here for consistency.

File: core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt
Comment (fire-at-will) [reply core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt:15]: Added in https://github.com/RevenueCat/purchases-kmp/pull/139/commits/130f23b825dc7de8ffedd7670d78040b415b0443!

File: core/src/iosMain/kotlin/com/revenuecat/purchases/kmp/models/StoreTransaction.ios.kt
Comment (fire-at-will) [reply core/src/iosMain/kotlin/com/revenuecat/purchases/kmp/models/StoreTransaction.ios.kt]: I love this! TIL about the `Result` class - it's pretty cool! It's pretty similar to Swift's [Result](https://developer.apple.com/documentation/swift/result), the biggest difference seems to be that Swift's is enum-based. Thanks for showing me something new! 😄

I've changed StoreTransaction.fromMap to return a `Result` in https://github.com/RevenueCat/purchases-kmp/pull/139/commits/75c465bfaf8b7fa24016864d2a3fe378167cf37a

File: core/src/commonTest/kotlin/com/revenuecat/purchases/kmp/PurchasesAreCompletedByTests.kt
Comment (fire-at-will) [reply core/src/commonTest/kotlin/com/revenuecat/purchases/kmp/PurchasesAreCompletedByTests.kt]: Hah, old habits die hard, don't they? Plain English function names for tests is _such a cool idea_, all languages should totally steal this feature 😄

I've updated all of the tests to use sentences with backticks in https://github.com/RevenueCat/purchases-kmp/pull/139/commits/9f36053dec99e25235c7e2f850b4bad12356d904.

File: core/src/iosMain/kotlin/com/revenuecat/purchases/kmp/Purchases.ios.kt
Comment (fire-at-will) [reply core/src/iosMain/kotlin/com/revenuecat/purchases/kmp/Purchases.ios.kt:23]: I'm honestly not sure how these got added in - I'm using the most vanilla installation of Android Studio imaginable 😅 Maybe I opened the file in VSCode and that reformatted it? Either way, I'll see if I can revert the styling changes.

File: .github/workflows/main.yml
Comment (fire-at-will) [reply .github/workflows/main.yml:59]: Sounds good! When you migrate it to CircleCI later, don't forget that `./gradlew test` only runs the JVM tests and we'll need to figure something else out for the iOS tests 😅

File: core/src/iosMain/kotlin/com/revenuecat/purchases/kmp/Purchases.ios.kt
Comment (fire-at-will) [reply core/src/iosMain/kotlin/com/revenuecat/purchases/kmp/Purchases.ios.kt:23]: Okay, I think I got them all reverted! 😄

File: .github/workflows/main.yml
Comment (JayShortway) [reply .github/workflows/main.yml:59]: Yeah on that topic, just realized today that `./gradlew allTests` should also run the iOS tests 😄

File: apiTester/src/androidMain/kotlin/com/revenuecat/purchases/kmp/apitester/PurchasesAreCompletedByAPI.kt
Comment (JayShortway) [reply apiTester/src/androidMain/kotlin/com/revenuecat/purchases/kmp/apitester/PurchasesAreCompletedByAPI.kt]: Nice! Could you also move the `PurchasesAreCompletedByAPI.kt` file to `commonMain`? 🙏

File: core/src/androidMain/kotlin/com/revenuecat/purchases/kmp/Purchases.android.kt
Comment (JayShortway) [reply core/src/androidMain/kotlin/com/revenuecat/purchases/kmp/Purchases.android.kt]: Good call! I added [SDK-3554](https://linear.app/revenuecat/issue/SDK-3554/standardize-handling-of-functions-unsupported-on-a-single-platform).

File: core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesAreCompletedBy.kt
Comment (JayShortway) [reply core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesAreCompletedBy.kt]: Even better 😄

File: core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt
Comment (JayShortway) [reply core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt]: Ah thanks for explaining! What do you think about doing this 👇  _only_ for purchases-kmp?
>we might could have simplified this by adding `storeKitVersion` to `PurchasesAreCompletedBy.RevenueCat` and forgoing the `storeKitVersion` parameter entirely

At first my opinion was that it's a no-go for consistency, but then I thought that possibly very few people will be using multiple RevenueCat SDKs, and comparing them. So curious what you think?

Comment (JayShortway) [reply core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt]: Regarding the tag for logging: it's a convention to use the class name as a tag. So in this case that would be `PurchasesConfiguration`. **However**, I see that [in purchases-android we use `[Purchases]`](https://github.com/RevenueCat/purchases-android/blob/main/purchases/src/main/kotlin/com/revenuecat/purchases/common/logUtils.kt#L48), so I'd say let's use that here too. 😄

File: core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt
Comment (JayShortway) [reply core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt]: Got it, thanks! I do think that passing `null` could signal the same intent. ("I don't have an appropriate value for `storeKitVersion`. RC decides.") I'm fine leaving `DEFAULT` in though.

File: core/src/iosMain/kotlin/com/revenuecat/purchases/kmp/models/StoreTransaction.ios.kt
Comment (JayShortway) [reply core/src/iosMain/kotlin/com/revenuecat/purchases/kmp/models/StoreTransaction.ios.kt]: Yeah it's pretty similar indeed! Would've loved for Kotlin's to also have typed failures. On the other hand, if the need for that gets stronger, it's very easy to build something ourselves. 
  
(In my own apps I use the third-party [`Either`](https://apidocs.arrow-kt.io/arrow-core/arrow.core/-either/index.html), which is also why I added [an add-on for that](https://github.com/RevenueCat/purchases-kmp/blob/main/either/src/commonMain/kotlin/com/revenuecat/purchases/kmp/either/Either.kt) to KobanKat/purchases-kmp.)

File: core/src/commonTest/kotlin/com/revenuecat/purchases/kmp/PurchasesAreCompletedByTests.kt
Comment (JayShortway) [reply core/src/commonTest/kotlin/com/revenuecat/purchases/kmp/PurchasesAreCompletedByTests.kt]: Haha yeah, totally understandable! 

>all languages should totally steal this feature 😄

I recently learned that Ruby (Spec) has it too!

File: core/src/iosMain/kotlin/com/revenuecat/purchases/kmp/Purchases.ios.kt
Comment (JayShortway) [reply core/src/iosMain/kotlin/com/revenuecat/purchases/kmp/Purchases.ios.kt:23]: VSCode might have done it indeed 🤔 

You can auto-format using CMD+OPTION+L. You were nearly there haha. Did the rest here: https://github.com/RevenueCat/purchases-kmp/pull/149.

File: apiTester/src/androidMain/kotlin/com/revenuecat/purchases/kmp/apitester/PurchasesAreCompletedByAPI.kt
Comment (fire-at-will) [reply apiTester/src/androidMain/kotlin/com/revenuecat/purchases/kmp/apitester/PurchasesAreCompletedByAPI.kt]: Yes! Done in https://github.com/RevenueCat/purchases-kmp/pull/139/commits/6e40eb030bf798fcf707f9f0f9e4ba8f44d9a104

File: core/src/iosMain/kotlin/com/revenuecat/purchases/kmp/Purchases.ios.kt
Comment (fire-at-will) [reply core/src/iosMain/kotlin/com/revenuecat/purchases/kmp/Purchases.ios.kt]: Merged!

File: .github/workflows/main.yml
Comment (fire-at-will) [reply .github/workflows/main.yml:59]: Oh nice! I changed the CI additions in this PR to run `./gradlew allTests` instead of `./gradlew test` here: https://github.com/RevenueCat/purchases-kmp/pull/139/commits/f83eaac0f5ced3d675f674d252eebfb19f4b752a

File: core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt
Comment (fire-at-will) [reply core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt]: I think my personal preference would be to keep it as is for the sake of API consistency across the hybrids. It's true that developers shouldn't be using multiple SDKs, but I'm afraid that allowing the APIs to drift apart could cause future development across the different hybrids to be more difficult.

File: core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/Purchases.kt
Comment (MarkVillacampa) [inline core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/Purchases.kt:100]: This is so `purchasesAreCompletedBy` cannot be changed after configuration, right? If so 👍

File: core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt
Comment (MarkVillacampa) [inline core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt:55]: Nice messages! 👍

**Review by MarkVillacampa — APPROVED** [review]

**Review by JayShortway — APPROVED** [review]
Great job! Looks 😚👌

File: core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/Purchases.kt
Comment (fire-at-will) [reply core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/Purchases.kt:100]: Yeah! I couldn't think of a good use case to change it, so we decided to remove it.

File: apiTester/src/commonMain/kotlin/com/revenuecat/purchases/kmp/apitester/PurchasesAreCompletedByAPI.kt
Comment (tonidero) [inline apiTester/src/commonMain/kotlin/com/revenuecat/purchases/kmp/apitester/PurchasesAreCompletedByAPI.kt:12]: Could be good to assign to a variable to verify the type:
```suggestion
                val storeKitVersion: StoreKitVersion  = mode.storeKitVersion
```

File: core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesAreCompletedBy.kt
Comment (tonidero) [inline core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesAreCompletedBy.kt:12]: Hmm does this need to be a `data object`? I think it could just be an object, but lmk if I missed something!

File: core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt
Comment (tonidero) [inline core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt:15]: Hmm I would make this nullable, like the `appUserId` or `userDefaultsSuiteName`. This way, we don't need to be duplicating the defaults here.

File: core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt
Comment (tonidero) [inline core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt:28]: Should we also add the `pendingTransactionsForPrepaidPlansEnabled` option here? We can also do it in a minor later, but it would be great to do it in the same version we bump the native.

File: core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt
Comment (tonidero) [inline core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt:44]: Been thinking about logs in the hybrids for a while. We don't really have a great way of exposing these... So I think the solution in this PR works for now. But maybe we should create a top-level accessor for the `Purchases.logHandler`? Something like `fun handleLog(level, message)` that internally calls `Purchases.logHandler`...

In any case, I think this comment can be addressed separately cc @JayShortway

**Review by tonidero — COMMENTED** [review]
Looks like we haven't added the `installmentsInfo` to the `SubscriptionOption` in Android either right? Again, I don't think it's a blocker, but ideally we also add that property there.

File: apiTester/src/commonMain/kotlin/com/revenuecat/purchases/kmp/apitester/PurchasesAreCompletedByAPI.kt
Comment (fire-at-will) [reply apiTester/src/commonMain/kotlin/com/revenuecat/purchases/kmp/apitester/PurchasesAreCompletedByAPI.kt]: Love this! I applied your suggestion in commit https://github.com/RevenueCat/purchases-kmp/pull/139/commits/ae097bbafd6abb97346c7b83be07062b1e58e351

File: core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesAreCompletedBy.kt
Comment (JayShortway) [reply core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesAreCompletedBy.kt:12]: `data object` is the recommended type in `sealed` hierarchies, with a minor lint "suggestion" appearing if you omit it: 
Scherm­afbeelding 2024-08-19 om 15 25 05

Reasoning in [the docs](https://kotlinlang.org/docs/object-declarations.html#using-data-objects-with-sealed-hierarchies):
> [...] they allow you to maintain symmetry with any data classes you may have defined alongside the object. In this example, declaring `EndOfFile` as a `data object` instead of a plain `object` means that it will get the `toString()` function without the need to override it manually:

File: core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt
Comment (fire-at-will) [reply core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt:15]: It looks like `IosPurchases.configureWithAPIKey()` requires a non-null `String` to be passed in, so I think it's probably best to keep it as is for now. The default value is `StoreKitVersion.DEFAULT`, which the native iOS SDK is responsible for resolving, so in a roundabout way I guess we're still not duplicating the default version here 😅

File: core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt
Comment (JayShortway) [reply core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt:44]: Just curious, but what would be the advantage of a `handleLog()` function?

File: core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt
Comment (JayShortway) [reply core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt:28]: I think we should add those new BC7 APIs before the v1 stable if possible. That way we can immediately advertise "full BC7 support". Doesn't need to be in this PR though, but can be.

File: core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesAreCompletedBy.kt
Comment (tonidero) [reply core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesAreCompletedBy.kt:12]: Makes sense, thanks!

**Comment by fire-at-will** [comment]
@tonidero About `SubscriptionOption.installmentInfo` - it looks like the Android `SubscriptionInfo` object is actually a typealias for the Android `com.revenuecat.purchases.models.SubscriptionOption`, which already has `installmentInfo` added to it 😄
Screenshot 2024-08-19 at 8 52 25 AM
Screenshot 2024-08-19 at 8 52 37 AM

File: core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt
Comment (JayShortway) [reply core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt:29]: Nice! What do you think about adding `installmentsInfo` (see Toni's parent comment)?

**Comment by JayShortway** [comment]
@tonidero @fire-at-will Yea that's true, but this means it's only usable in the Android source set. I think it's beneficial if we can access it from the common source set, so for that we'd need to add a public API in `commonMain/[...]/SubscriptionOption.kt`. However, we can also do that when we converting away from `typealias`es (but we should make sure not to forget).

**Comment by fire-at-will** [comment]
@JayShortway I'll give it a shot! Will be back in a few minutes! :D

File: core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt
Comment (tonidero) [reply core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt:15]: Oh right, looks like we're using the native configure method for iOS, instead of the PHC one, and that one requires it to be non-null... I guess it's ok then yeah!

**Comment by fire-at-will** [comment]
@JayShortway @tonidero The changes to add `InstallmentInfos` started to be spread across several files, so I opened a PR into this branch to make it easier to review: https://github.com/RevenueCat/purchases-kmp/pull/150

File: core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt
Comment (tonidero) [reply core/src/commonMain/kotlin/com/revenuecat/purchases/kmp/PurchasesConfiguration.kt:46]: Mostly having an indirection level lower than `Purchases.logHandler`... honestly it's nitpicking though, so I think it's also ok to leave as is!

**Review by tonidero — APPROVED** [review]
LGTM after also merging the installmentsInfo PR

**Pull request merged**
